manager/orchestrator/replicatedjob/orchestrator.go

```diff
@@ -45,6 +52,166 @@ func (o *Orchestrator) Run(ctx context.Context) {
+
+		// getting tasks with FindTasks should only return an error if we've
+		// made a mistake coding; there's no user-input or even reasonable
+		// system state that can cause it. If it returns an error, we'll just
+		// panic and crash.
+		var err error
+		tasks, err = store.FindTasks(tx, store.ByServiceID(id))
+		if err != nil {
+			panic(fmt.Sprintf("error getting tasks: %v", err))
+		}
+	})
+
+	if service == nil {
+		return nil
+	}
+
+	// if this is the first iteration of the service, it may not yet have a
+	// JobStatus, so we should create one if so. this won't actually be
+	// committed, though.
+	if service.JobStatus == nil {
+		service.JobStatus = &api.JobStatus{}
+	}
+
+	// Jobs can be run in multiple iterations. The JobStatus of the service
+	// indicates which Version of iteration we're on. We should only be looking
+	// at tasks of the latest Version
+
+	jobVersion := service.JobStatus.JobIteration.Index
+
+	// now, check how many tasks we need and how many we have running. some of
+	// these tasks may be on their way to completion, but that just means we
+	// won't start enough tasks, which isn't a problem. also, we don't care if
+	// tasks are failed here. failed tasks don't factor into our computation of
+	// how many new tasks to create.
+	//
+	// also, for the math later, we need these values to be of type uint64.
+	runningTasks := uint64(0)
+	completeTasks := uint64(0)
+	// slots keeps track of the slot numbers available for tasks. as long as
+	// the orchestrator isn't broken, then this map will consist of the set of
+	// all integers from 0 to (MaxConcurrent - 1), with a boolean indicating if
+	// that slot is occupied. This slot handling code is much simpler than the
+	// code for replicated services, because we don't need to worry about
+	// restarts.
+	slots := map[uint64]bool{}
+	for _, task := range tasks {
+		// we only care about tasks from this job iteration. tasks from the
+		// previous job iteration are not important
+		// TODO(dperny): we need to stop any running tasks from older job
+		// iterations.
+		if task.JobIteration != nil && task.JobIteration.Index == jobVersion {
+			if task.Status.State == api.TaskStateCompleted {
+				completeTasks++
+			}
+
+			// any tasks that are created and running but not yet terminal are
+			// considered "running tasks" for our purpose, because they don't
+			// need to be created.
+			if task.Status.State <= api.TaskStateRunning && task.DesiredState == api.TaskStateCompleted {
+				runningTasks++
+				slots[task.Slot] = true
+			}
+		}
+	}
+
+	// now that we have our counts, we need to see how many new tasks to
+	// create. this number can never exceed MaxConcurrent, but also should not
+	// result in us exceeding TotalCompletions. first, get these numbers out of
+	// the service spec.
+	rj := service.Spec.GetReplicatedJob()
+
+	// possibleNewTasks gives us the upper bound for how many tasks we'll
+	// create. also, ugh, subtracting uints. there's no way this can ever go
```

Comment:
`subtracting uints` => was just about to say that :-P might be worth checking that `rj.MaxConcurrent > runningTasks ` first?

Reply:
It's a bit oddly constructed, but basically, the variable `actualNewTasks` can take one of two values: that of `possibleNewTasks` and that of `allowedNewTasks`, each of which is not dependent on the other.  Before creating any tasks, we check if `actualNewTasks` is larger than `TotalCompletions`, and if so, we know there has been some kind of underflow in either case. This covers both cases with one easily-testable swoop.